Implement source-aware updates to speed up dependency checks

Arcitec opened the pull request:
OneTrainer's self-update process now supports source-aware updates via an opt-in flag, which skips the update check if the OneTrainer source code hasn't changed.

This optimization significantly speeds up executions of `update.sh`.

While generally safe, this approach may miss some updates and critical bug fixes for third-party dependencies.

By default, this feature is disabled to ensure maximum safety.

Closes: #637

Arcitec commented:
@Nerogar Hey, hope the new year has been treating you well! :) I am currently very busy in my spare time, and had to make the time to implement this feature which @dxqbYD 's requested for the cloud training feature. It will significantly speed up the RunPod startup, since the cloud code runs `update.sh` on every training startup.

Everything has been tested deeply and implemented cleanly within the current framework. I don't really have time for revisions and small comment changes and things like this, and it may take me days to notice any comments here since I am swamped with work, so I hope that everything here is to your satisfaction and can be merged. :+1: :heart:

dxqbYD reviewed (commented):
@Arcitec thanks for finding the time anyway! The new feature works for me.

Final review is up to @Nerogar but since I have initiated this feature I've left some comments anyway.

Arcitec commented on LAUNCH-SCRIPTS.md:
```diff
@@ -23,6 +23,8 @@
 
 - `OT_PREFER_VENV`: If set to `true`, Conda will be ignored even if it exists on the system, and Python Venv will be used instead. This ensures that people who use `pyenv` (to choose which Python version to run on the host) can easily set up their desired Python Venv environments. Defaults to `false`.
 
+- `OT_LAZY_UPDATES`: If set to `true`, OneTrainer's self-update process will only update the Python environment's dependencies if the OneTrainer source code has been modified since the previous dependency update. This speeds up executions of `update.sh`, and is generally safe, but may miss some updates and important bugfixes for external third-party dependencies. If you use this option, you must  ...
```
The risk is not in our direct dependencies. It's in our direct dependencies' dependencies. Let's say we use `diffusers==1.0`. Well, they have looser requirements, like this:

https://github.com/huggingface/diffusers/blob/main/src/diffusers/dependency_versions_table.py

Examples:

- `"accelerate>=0.31.0",`
- `"huggingface-hub>=0.27.0",`
- etc.

This is true for most of our dependencies. They aren't as strict as us.

Now think of the scenario where there's a bug in `huggingface-hub 0.27.0` but it's fixed in `0.27.1`.

With `./update.sh` we will update to the fixed version because our dependency graph will resolve that dependency to the latest available version.

With `OT_LAZY_UPDATES="true" ./update.sh` we will not get that fix.

So yes it's a real problem, and it's impossible to guess how big the bugs may be. It can be anything from small bugs to critical fixes that are being missed by not doing a full update. That's why it still does full non-"lazy" updates by default. :)

dxqbYD replied:
I don't think pip would detect that even in "non-lazy" mode, would it?
If OT depends on A=1.0 but A depends on B>=2.0 and B releases a version 2.1, I don't think pip would do anything when you run pip install -r requirements.txt on OT, which only depends on A=1.0. The dependency is already fulfilled, no action either way, unless you have implemented special handling of these cases.

Arcitec replied:
Historically, pip always upgraded to the newest versions that still satisfy the requirements.

But since many people install dependencies manually via `pip install numpy==1.0; pip install diffusers` etc, this led to users breaking their environments, because later installs will upgrade things that were manually installed earlier.

So around 10 years ago, the decision was made to make pip "lazy by default". Nowadays, it keeps the currently installed package version, or installs outdated locally cached downloaded packages on fresh `pip install` calls, as long as what it already has locally satisfies the general requirement version range. This means that "pip install" by default doesn't upgrade anything, in most cases. Not even during fresh installs (if there's an older cached version of a package).

(Other Python package managers don't have this problem, by the way. They usually keep a "pinned packages" history of exactly what has been manually installed in the past, and will update everything to the latest versions that satisfies all old, manual installs. But since pip is a basic package manager, they had to make their defaults reasonable for beginners.)

To still allow people to install actual package updates, they also added a custom `--upgrade-strategy eager` flag, for when you want to actually upgrade the tree to the best possible state and know what you're doing. The correct way to use that flag is to provide it *and* a list of *all* dependencies and requirements.txt files at the same time, in a *single* command, so that the resolver can check the whole graph and get the best versions of everything, while always ensuring that every package has been satisfied. (The only thing that should be done separately is upgrades of core package tools: pip and setuptools itself).

That's the strategy we use. You can find it with `--upgrade-strategy eager` in `lib.include.sh`. And the same fix has been brought over to the Windows update.bat too.

So when users run our `./update.sh`, they get the freshest, best possible, most-fixed versions of all dependencies, while still ensuring that the whole dependency graph is happy. It's a true update. Which means that when bugs are fixed in dependencies of *our* dependencies, we get those fixes when we update. :)

Arcitec commented on lib.include.sh:
```diff
@@ -77,6 +79,36 @@ function print_command {
+# Retrieves the most recent Git commit's hash.
+function get_current_git_hash {
+    # NOTE: Will not detect local code changes, only the newest commit's hash,
+    # but we append a trailing "*" asterisk in case of local changes.
```
Yeah it's unfortunate that it only detects the first change, I know. I figured that it was better than no detection of code changes at all. But perhaps you are right that we should treat uncommitted code changes as "nothing has been changed" in terms of update behavior. Developers can run updates without `OT_LAZY_UPDATES` to force updates in that case.

I am going to change that behavior to use the latest commit hash at all times.

Arcitec commented on lib.include.sh:
```diff
@@ -77,6 +79,36 @@ function print_command {
+    # but we append a trailing "*" asterisk in case of local changes.
+    # NOTE: We tell it to only output tag names if they match "NO TAG" (never
+    # matches), which forces it to always output the HASH instead of tag names.
+    git describe --always --abbrev=0 --match "NO TAG" --dirty="*"
```
Both are totally reliable, but since I agree with you about removing the `*` asterisk for "dirty git workdirs", there's no need for this alternative method (since it's only needed for implementing "dirty source detection").

I am going to switch to `git rev-parse HEAD`.

Arcitec commented on lib.include.sh:
```diff
@@ -25,6 +25,7 @@ export OT_CONDA_ENV="${OT_CONDA_ENV:-conda_env}"
 export OT_PYTHON_CMD="${OT_PYTHON_CMD:-python}"
 export OT_PYTHON_VENV="${OT_PYTHON_VENV:-venv}"
 export OT_PREFER_VENV="${OT_PREFER_VENV:-false}"
+export OT_LAZY_UPDATES="${OT_LAZY_UPDATES:-false}"
```
Yes only `true` works, and that's how every true/false parameter in OneTrainer's launcher behaves.

The value was chosen because because `0` and `1` do not act properly in Bash. It's a bit of a stupid scripting language and pure numbers are buggy in that language. Take this example:

```
[[ "hi" -eq 0 ]] && echo "hi"
Result: hi
```

So it thinks "hi" is equal to 0.

This is one reason why it's preferable to write a full word such as "true"/"false" or "yes"/"no" in environment variables. There's no risk of Bash interpreting those incorrectly.

It's also nicer and clearer to look at in the library source code, but it does mean that you need to write out the full word "true" (or "false") in those cases, which is definitely slightly inconvenient, but it's more reliable that way.

Arcitec commented:
@dxqbYD No worries. I spotted the email about your code review now while dealing with something else, and I agree with you that we might as well remove the "dirty git workdir" detection code, and treat modified worktrees as "no changes since the last commit".

My original reasoning was mainly that we're not "really" on the latest commit anymore after we have changed the tracked files, which is why I made it append an asterisk for modified code. Technically things have changed. But since it's impossible to detect what has changed (or when multiple changes happen later), I agree with you that it doesn't add much to the robustness of the git commit detection.

So I will be reverting to purely looking at the latest full commit hash now, without any "workdir change detection".

Arcitec commented:
@dxqbYD The other comment about gitignore was missing above (maybe deleted it?) so I forgot to do that change, but I've force-pushed that too now:

https://github.com/Nerogar/OneTrainer/compare/679e1faa918726680e32b5e51d87fc1a8a74ffc0..5dec1e98de90300ce232af2066543a5ef0a34596

It was a good suggestion. We now ignore all files ending in tilde, since some text editors keep "temporary text copies" in temp files ending in `~` as you pointed out, and we definitely don't want to be annoyed by those in `git status` or `git add .`! :)

dxqbYD commented:
> your code review now while dealing with something else, and I agree with you that we might as well remove the "dirty git workdir" detection code, and treat modified worktrees as "no changes since the last commit".
> 

I deleted it because I saw that you are busy, and it was an unrelated improvement that we don't need for this

dxqbYD reviewed (approved):
now used in https://github.com/dxqbYD/OneTrainer/tree/cloud_dev

Arcitec commented:
@dxqbYD Hey again! Nice to see it's merged and being used already. :)

Your usage here looks good:

https://github.com/dxqbYD/OneTrainer/commit/ac27025d3d5034fc8f0fe50e43f299ec58012f94

I should probably explain a bit more details about why the usage documentation says "If you use this option, you must set it permanently for *every* script (not just `update.sh`)."

That's just a simplified way to sum up the situation which is as follows:

- `update.sh`: It always checks the `OT_LAZY_UPDATES` variable on every run and uses it to determine whether to attempt lazy updates, or to do a full update and delete the "latest update git hash" metadata file. So yeah it always needs to be there.
- **All** other scripts: They check if the Python environment (venv/conda) exists. If it doesn't exist, they create a new environment and then write or delete the "latest update git hash" metadata file based on `OT_LAZY_UPDATES`. But if the environment already exists, `OT_LAZY_UPDATES` doesn't do anything at all.

That's why I summed it up as "you should set it permanently" so that people who use it get the expected results.

You can even see it if you look at the code for these scripts. They all run `prepare_runtime_environment` which verifies the environment, creates if missing, and activates it so that subsequent commands can run inside the environment. Install.sh is basically just a leftover/courtesy scripts for people who want to manually force an install ahead of time, without starting anything afterwards.

- https://github.com/Nerogar/OneTrainer/blob/master/install.sh
- https://github.com/Nerogar/OneTrainer/blob/master/start-ui.sh
- https://github.com/Nerogar/OneTrainer/blob/master/run-cmd.sh

But since your cloud runner manually starts `install.sh` and `update.sh`, it seems like you've already covered every situation where installs/updates will be happening. So you probably don't need to add it to `run-cmd.sh`. But it wouldn't be a bad idea to add it there too, in case you've missed a scenario where `run-cmd.sh` would detect that the environment is deleted and will create a new one. Just mentioning it to make you aware. Hope this extra info helps! :)
